client/my-sites/importer/importing-pane.jsx

```diff
@@ -14,6 +15,61 @@ import ProgressBar from 'components/progress-bar';
 import MappingPane from './author-mapping-pane';
 import Spinner from 'components/spinner';
 
+const sum = ( a, b ) => a + b;
+
+/*
+ * The progress object comes from the API and can
+ * contain different object counts.
+ *
+ * The attachments will lead the progress because
+ * they take the longest in almost all circumstances.
+ *
+ * progressObect ~= {
+ *     post: { completed: 3, total: 12 },
+ *     comment: { completed: 0, total: 3 },
+ *     …
+ * }
+ */
+const calculateProgress = progress => {
+	const { attachment = {} } = progress;
+
+	if ( attachment.total > 0 && attachment.completed >= 0 ) {
+		// return a weight of 80% attachment, 20% other objects
+		return 80 * attachment.completed / attachment.total +
+		       0.2 * calculateProgress( omit( progress, [ 'attachment' ] ) );
+	}
+
+	const percentages = Object.keys( progress )
+		.map( k => progress[ k ] ) // get the inner objects themselves
+		.filter( ( { total } ) => total > 0 ) // skip ones with no objects to import
+		.map( ( { completed, total } ) => completed / total ); // compute the individual percentages
+
+	return 100 * percentages.reduce( sum, 0 ) / percentages.length;
+};
+
+const resourcesRemaining = progress => Object.keys( progress )
+		.map( k => progress[ k ] )
+		.map( ( { completed, total } ) => total - completed )
+		.reduce( sum, 0 );
+
+const hasProgressInfo = progress => {
+	if ( ! progress ) {
+		return false;
+	}
+
+	const types = Object.keys( progress ).map( k => progress[ k ] );
+	if ( ! types.length ) {
+		return false;
+	}
+
+	const firstType = types.shift();
+	if ( ! firstType.completed ) {
```

Comment:
This is a trickier one. The API currently _may_ or _may not_ send that field on the object. The goal of this code is to determine if indeed we have the values we expect. Now, strangely enough, when I tried either of the next two methods, I introduced a bug:

```js
// method 1: lodash
if ( ! has( firstType, 'completed' ) )

// method 2: hasOwnProperty
if ( ! firstType.hasOwnProperty( 'completed' ) )
```

The bug was that before we started tracking progress this function apparently returned true and the "Finishing up the import" message displayed. So far, I haven't dived into figuring out why.